llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h

```diff
@@ -491,6 +485,15 @@ m_Select(const Op0_t &Op0, const Op1_t &Op1, const Op2_t &Op2) {
       {Op0, Op1, Op2});
 }
 
+template <typename Op0_t>
+inline match_combine_or<UnaryVPInstruction_match<Op0_t, VPInstruction::Not>,
+                        AllBinaryRecipe_match<int_pred_ty<is_all_ones>, Op0_t,
+                                              Instruction::Xor, true>>
+m_Not(const Op0_t &Op0) {
```

Comment:
It might be better to canonicalize negation via `xor` to `Not`, although it might cause cost-model issues?

Reply:
Yes I had the same thought, or just to remove VPInstruction::Not entirely. I can explore this in a separate PR?

Reply:
This can be done separately, yes. It's probably preferable to keep `VPInstruction::Not`, as it's clearer and doesn't require creating a constant.